tf2onnx/onnx_opset/logical.py

```diff
@@ -29,16 +29,21 @@ def _add_cast_to_inputs(graph, node, supported_dtypes, target_dtype):
             graph.copy_shape(inp, inp_cast.output[0])
             graph.set_dtype(inp_cast.output[0], target_dtype)
 
-
-def _add_cast_to_same_type_to_inputs(graph, node):
+def _add_cast_to_same_type_to_inputs(graph, node, supported_dtypes, target_dtype):
     common_dtype = graph.get_dtype(node.input[0])
+    if common_dtype not in supported_dtypes:
```

Comment:
My bad, I misunderstood the issue before.

Actually, the issue is: The op Equal doesn't support 'string' with version 11, meaning we should fail the conversion once we detect this case. Even the content is empty, we should not convert it to an integer value which might confuse users.

So it'd better to set up an unsupported op list which may only contain string right now. If we detect current type of input[0] is in it, we fail the conversion with a reasonable message instead of making tricky things to work around it.

Make sense?

Reply:
> Actually, the issue is: The op Equal doesn't support 'string' with version 11, meaning we should fail the conversion once we detect this case.

In the OpSet 7 version of this, the Equal Op works fine assuming the strings can be converted to the desired type. So this change makes OpSet11 match the OpSet7 behavior. 

So if we made this an error, then OpSet 7 Equal and Opset 11 would behave differently but both have the same stance on strings. Making them behave different didn't seem right. Rather than remove what was working in OpSet 7, we decided to enable the same setup for OpSet 11. It could be argued to make OpSet7 (and earlier) also automatically error in this case instead. If that's truely preferred we could look into it. However our thought was to expand 11 rather than restrict 7.

> Even the content is empty, we should not convert it to an integer value which might confuse users.

I'll agree that the empty string was a special case that almost lead us to restrict OpSet 7. However after digging into the TF code, we found they explicitly handle this case. When creating a feature column, if an entry is `""` then it's explicitly changed to `-1` (https://github.com/tensorflow/tensorflow/blob/4e7f0185c70faf35e12acbfe381a729d1e6cc38c/tensorflow/python/feature_column/feature_column.py#L2286). Since it's explicitly handled, we matched the TF behavior over just erroring out. Otherwise it is also confusing to have a model work in TF but fail to convert.

For reference we have a model encountering this scenario that indeed works fine in TF but then fails to convert. With these changes, it converts fine. Unfortunately it's not a model we can share but it is a real world scenario.

Reply:
Thanks for your detailed explanations.